test/e2e/daemon_restart.go

```diff
@@ -0,0 +1,194 @@
+	var newPods cache.Store
+	var stopCh chan struct{}
+
+	BeforeEach(func() {
+
+		// These tests require SSH
+		if !providerIs("gce") {
+			By(fmt.Sprintf("Skipping test, which is not implemented for %s", testContext.Provider))
+			return
+		}
+		framework.beforeEach()
+		ns = framework.Namespace.Name
+
+		// All the restart tests need an rc and a watch on pods of the rc.
+		// Additionally some of them might scale the rc during the test.
+		config = RCConfig{
+			Client:      framework.Client,
+			Name:        rcName,
+			Namespace:   ns,
+			Image:       "kubernetes/pause",
+			Replicas:    numPods,
+			CreatedPods: &[]*api.Pod{},
+		}
+		Expect(RunRC(config)).NotTo(HaveOccurred())
+		storePods(*config.CreatedPods, existingPods)
+
+		stopCh = make(chan struct{})
+		newPods, controller = controllerFramework.NewInformer(
+			&cache.ListWatch{
+				ListFunc: func() (runtime.Object, error) {
+					return framework.Client.Pods(ns).List(labelSelector, fields.Everything())
+				},
+				WatchFunc: func(rv string) (watch.Interface, error) {
+					return framework.Client.Pods(ns).Watch(labelSelector, fields.Everything(), rv)
+				},
+			},
+			&api.Pod{},
+			0,
+			controllerFramework.ResourceEventHandlerFuncs{},
+		)
+		go controller.Run(stopCh)
+	})
+
+	AfterEach(func() {
+		close(stopCh)
+		expectNoError(DeleteRC(framework.Client, ns, rcName))
+		framework.afterEach()
+	})
+
+	It("Controller Manager should not create/delete replicas across restart", func() {
+
+		restarter := NewRestartConfig(
+			getMasterHost(), "kube-controller", ports.ControllerManagerPort, restartPollInterval, restartTimeout)
+		restarter.Restart()
+
+		// The intent here is to make sure the rc we created has synced at least once post restart before
+		// validating that it didn't create/delete pods. Scaling the rc to the same size achieves this,
+		// because the scale operation will wait on the sequence number of the status to match spec.
+		ScaleRC(framework.Client, ns, rcName, numPods, true)
+
+		// Only check the keys, the pods can be different if the kubelet updated it.
+		// TODO: Can it really?
+		existingKeys := util.NewStringSet()
+		newKeys := util.NewStringSet()
+		for _, k := range existingPods.ListKeys() {
+			existingKeys.Insert(k)
+		}
+		for _, k := range newPods.ListKeys() {
+			newKeys.Insert(k)
+		}
+		if len(newKeys.List()) != len(existingKeys.List()) ||
+			!newKeys.IsSuperset(existingKeys) {
+			Failf("RcManager created/deleted pods after restart")
+		}
+	})
+
+	It("Scheduler should continue assigning pods to nodes across restart", func() {
+
+		restarter := NewRestartConfig(
+			getMasterHost(), "kube-scheduler", ports.SchedulerPort, restartPollInterval, restartTimeout)
+
+		// Create pods while the scheduler is down and make sure the scheduler picks them up by
+		// scaling the rc to the same size.
+		restarter.waitUp()
+		restarter.kill()
+		expectNoError(ScaleRC(framework.Client, ns, rcName, numPods+5, false))
```

Comment:
This should be good enough, IIUC how Monit works. IIUC monit does not restart things straight away, but marks daemons as unresponsive and restarts them in the next pass (which is 2 minutes later). If I'm mistaken and Monit tries to restart things instantly, then we can get false positives here. Either way it needs a comment.